### PR TITLE
Static class support (revised)

### PR DESCRIPTION
**Pull request by moodmosaic**
Revised version of #18 based on the comments of #19.

### CODE REVIEW

**Comment by ploeh** [comment]
Thank you for your contribution. I'm wondering if the algorithm chosen is the simplest thing that could possibly work.

Given that we have two overloads of the `TestCase` constructor:

``` c#
TestCase(Action)
TestCase(Action<object>)
```

isn't it clear which one needs an instance of the test class, and which one doesn't?

**Comment by moodmosaic** [comment]
Do you mean that, instead of adding an argument in the constructor, we could probably make the _ShouldCreateInstance_ property more intelligent by inspecting the Action delegate?

**Comment by moodmosaic** [comment]
If we do the following we won't need the additional boolean constructor argument:

```
public override bool ShouldCreateInstance
{
    get
    {
        return !this.testMethod.Class.IsAbstract 
            && !this.testMethod.Class.IsSealed;
    }
}
```

all tests pass, even those in the static class. What do you think? :)

**Comment by ploeh** [comment]
When an instance of `TestCase` is created, `TaskCase` _knows_ which constructor overload was called...

**Comment by moodmosaic** [comment]
I am lost, sorry :$

**Comment by ploeh** [comment]
When you invoke `TestCase(Action)`, you implicitly state that you don't care about any test class instance, because the signature of `Action` is `void Action()`; there's no input parameter. This means that already when that constructor overload is being invoked, `TestCase` _knows_ that when `ConvertToTestCommand` is invoked, it should pass `false` as the `shouldCreateInstance` flag to `FirstClassCommand`.

Likewise, when you invoke `TestCase(Action<object>)`, you implicitly state that you require the test class instance, because the signature of `Action<object>` is `void Action(object)`. The input parameter represents the test class instance. Again, when that constructor overload is used, `TestCase` _knows_ that `shouldCreateInstance` should be `true`.

**Comment by moodmosaic** [comment]
Thank you for your thorough explanation! I just added a [commit](https://github.com/moodmosaic/Exude/commit/04e7473f756c385637e9fe6134f49f21e1132f77) which, hopefully, implements in code what you described.

File: Src/Exude/FirstClassTestsAttribute.cs [hunk 1]
```diff
@@ -91,6 +90,14 @@ private static bool IsReturnTypeInvalid(IMethodInfo method)
         {
             return !typeof(IEnumerable<ITestCase>).IsAssignableFrom(
                 method.MethodInfo.ReturnType);
+        }
+
+        private static object GetReturnValue(IMethodInfo method)
```
Comment (ploeh) [inline Src/Exude/FirstClassTestsAttribute.cs:95]: Do we need this helper method? What happens if we delete it?

File: Src/Exude/FirstClassTestsAttribute.cs [hunk 2]
```diff
@@ -91,6 +90,14 @@ private static bool IsReturnTypeInvalid(IMethodInfo method)
...
             return !typeof(IEnumerable<ITestCase>).IsAssignableFrom(
                 method.MethodInfo.ReturnType);
+        }
+
+        private static object GetReturnValue(IMethodInfo method)
+        {
+            if (method.Class.IsAbstract && method.Class.IsSealed)
+                return method.MethodInfo.Invoke(null, null);
+
+            return method.MethodInfo.Invoke(method.CreateInstance(), null);
```
Comment (moodmosaic) [inline Src/Exude/FirstClassTestsAttribute.cs:100]: For static types we need to pass `null` while for non-static types we need to pass `method.CreateInstance()`.

**Comment by ploeh** [comment]
Thank you for your contribution! It's now live as [Exude 0.2.1](http://www.nuget.org/packages/Exude/0.2.1).

**Comment by moodmosaic** [comment]
Thank you for your help!

**Comment by ploeh** [comment]
Likewise :)

**Pull request merged**